source/extensions/filters/http/cache/http_cache.cc

```diff
@@ -0,0 +1,156 @@
+                                 "with null Path.");
+  ASSERT(request_headers.ForwardedProto(),
+         "Can't form cache lookup key for malformed Http::HeaderMap with null ForwardedProto.");
+  ASSERT(request_headers.Host(), "Can't form cache lookup key for malformed Http::HeaderMap "
+                                 "with null Host.");
+  const Http::HeaderString& forwarded_proto = request_headers.ForwardedProto()->value();
+  const auto& scheme_values = Http::Headers::get().SchemeValues;
+  ASSERT(forwarded_proto == scheme_values.Http || forwarded_proto == scheme_values.Https);
+  // TODO(toddmgreer): Let config determine whether to include forwarded_proto, host, and
+  // query params.
+  // TODO(toddmgreer): get cluster name.
+  // TODO(toddmgreer): Parse Range header into request_range_spec_, and handle the resultant
+  // vector<AdjustedByteRange> in CacheFilter::onOkHeaders.
+  key_.set_cluster_name("cluster_name_goes_here");
+  key_.set_host(std::string(request_headers.Host()->value().getStringView()));
+  key_.set_path(std::string(request_headers.Path()->value().getStringView()));
+  key_.set_clear_http(forwarded_proto == scheme_values.Http);
+}
+
+// Unless this API is still alpha, calls to stableHashKey() must always return
+// the same result, or a way must be provided to deal with a complete cache
+// flush. localHashKey however, can be changed at will.
+size_t stableHashKey(const Key& key) { return MessageUtil::hash(key); }
+size_t localHashKey(const Key& key) { return stableHashKey(key); }
+
+// Returns true if response_headers is fresh.
+bool LookupRequest::isFresh(const Http::HeaderMap& response_headers) const {
+  if (!response_headers.Date()) {
+    return false;
+  }
+  const Http::HeaderEntry* cache_control_header = response_headers.CacheControl();
+  if (cache_control_header) {
+    const SystemTime::duration effective_max_age =
+        Utils::effectiveMaxAge(cache_control_header->value().getStringView());
+    return timestamp_ - Utils::httpTime(response_headers.Date()) < effective_max_age;
+  }
+  // We didn't find a cache-control header with enough info to determine
+  // freshness, so fall back to the expires header.
+  return timestamp_ <= Utils::httpTime(response_headers.get(Http::Headers::get().Expires));
+}
+
+LookupResult LookupRequest::makeLookupResult(Http::HeaderMapPtr&& response_headers,
+                                             uint64_t content_length) const {
+  // TODO(toddmgreer): Implement all HTTP caching semantics.
+  ASSERT(response_headers);
+  LookupResult result;
+  result.cache_entry_status_ =
+      isFresh(*response_headers) ? CacheEntryStatus::Ok : CacheEntryStatus::RequiresValidation;
+  result.headers_ = std::move(response_headers);
+  result.content_length_ = content_length;
+  if (!adjustByteRangeSet(result.response_ranges_, request_range_spec_, content_length)) {
+    result.headers_->setStatus(416); // Range Not Satisfiable
```

Comment:
nit: define 416 in `Http::Codes` and use that?

Reply:
Done.